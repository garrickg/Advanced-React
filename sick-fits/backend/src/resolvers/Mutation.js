const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const { randomBytes } = require('crypto');
const { promisify } = require('util');
const { hasPermission } = require('../utils');

const { transport, makeANiceEmail } = require('../mail');

const ONE_YEAR = 1000 * 60 * 60 * 24 * 365;

const createToken = user => jwt.sign({ userId: user.id }, process.env.APP_SECRET);
const hashPassword = async password => bcrypt.hash(password, 10);

const Mutations = {
  async createItem(_parent, args, context, info) {
    const { userId } = context.request;
    // TODO: Check if user is authed
    if (!userId) {
      throw new Error('You must be logged in to add items!');
    }
    const item = await context.db.mutation.createItem(
      {
        data: {
          ...args,
          // this is how we create realtionships between the item and the user
          user: {
            connect: {
              id: userId,
            },
          },
        },
      },
      info,
    );
    return item;
  },
  updateItem(_parent, args, context, info) {
    // first take a copy of the udpates
    const updates = { ...args };
    // remove the ID from the udpates
    delete updates.id;
    // run the update method
    return context.db.mutation.updateItem(
      {
        data: updates,
        where: {
          id: args.id,
        },
      },
      info,
    );
  },
  async deleteItem(_parent, { id }, context, info) {
    const where = { id };
    // find the item
    const item = await context.db.query.item({ where }, '{ id, title, user { id } }');
    // TODO: check if they are the owner or admin
    const ownsItem = item.user.id === context.request.userId;
    const hasPermissions = context.request.user.permissions.some(permission => ['ADMIN', 'ITEMDELETE'].includes(permission));
    if (!ownsItem && !hasPermissions) {
      throw new Error('Insufficient permissions!');
    }
    // delete it
    return context.db.mutation.deleteItem({ where }, info);
  },
  async signup(_parent, args, context, info) {
    // lowercase the email
    const email = args.email.toLowerCase();
    // hash the password
    const password = await hashPassword(args.password);
    // create the user in the database
    const user = await context.db.mutation.createUser(
      {
        data: {
          ...args,
          email,
          password,
          permissions: { set: ['USER'] },
        },
      },
      info,
    );
    // create the JWT
    const token = createToken(user);
    // set JWT as cookie on the response
    context.response.cookie('token', token, {
      httpOnly: true,
      maxAge: ONE_YEAR, // 1 year cookie
    });
    return user;
  },
  async signin(_parent, { email, password }, context, _info) {
    const user = await context.db.query.user({ where: { email } });
    if (!user) {
      throw new Error('Incorrect email or password');
    }
    // check if password is correct
    const valid = await bcrypt.compare(password, user.password);
    if (!valid) {
      throw new Error('Incorrect email or password');
    }
    // generate jwt token
    const token = createToken(user);
    // set cookie with token
    context.response.cookie('token', token, {
      httpOnly: true,
      maxAge: ONE_YEAR,
    });
    return user;
    // return the user
  },
  signout(_parent, _args, context, _info) {
    context.response.clearCookie('token');
    return {
      message: 'Signed Out!',
    };
  },
  async requestReset(_parent, { email }, context, _info) {
    // check if user exists
    const user = await context.db.query.user({ where: { email } });
    // set a reset token and expiry for user
    if (!user) {
      throw new Error('Incorrect email');
    }
    const resetToken = (await promisify(randomBytes)(20)).toString('hex');
    const resetTokenExpiry = Date.now() + 3600000; // 1 hour from now
    const res = await context.db.mutation.updateUser({
      where: { email },
      data: { resetToken, resetTokenExpiry },
    });
    // email reset token
    const { FRONTEND_URL } = process.env;
    const mailRes = await transport.sendMail({
      from: 'garrickgunn@gmail.com',
      to: user.email,
      subject: 'Password Reset Link',
      html: makeANiceEmail(`Your Password Reset Token is here: \n\n <a href="${FRONTEND_URL}/reset?resetToken=${resetToken}">Click Here to Reset!</a>`),
    });
    // return the message
    return { message: 'reset requested' };
  },
  async resetPassword(_parent, { resetToken, password, confirmPassword }, context, _info) {
    // check if the passwords match
    if (password !== confirmPassword) {
      throw new Error('Passwords do not match');
    }
    // check if resetToken OK
    // check if token expired
    const [user] = await context.db.query.users({
      where: {
        resetToken,
        resetTokenExpiry_gte: Date.now(),
      },
    });
    if (!user) {
      throw new Error('Invalid token');
    }
    // hash new password
    const hashedPassword = await hashPassword(password);
    // save new password and remove token and expiry
    const updatedUser = await context.db.mutation.updateUser({
      where: { id: user.id },
      data: {
        password: hashedPassword,
        resetToken: null,
        resetTokenExpiry: null,
      },
    });
    // generate JWT
    const token = createToken(updatedUser);
    // set the JWT cookie
    context.response.cookie('token', token, {
      httpOnly: true,
      maxAge: ONE_YEAR,
    });
    // return the user
    return updatedUser;
  },
  async updatePermissions(parent, { permissions, userId: id }, context, info) {
    const { userId } = context.request;
    // check if logged in
    if (!userId) {
      throw new Error('Please login first!');
    }
    // query current user
    const currentUser = await context.db.query.user({
      where: {
        id: userId,
      },
    }, info);
    // check their permissions
    hasPermission(currentUser, ['ADMIN', 'PERMISSIONUPDATE']);
    // update the permissions
    return context.db.mutation.updateUser({
      data: {
        permissions: {
          set: permissions,
        },
      },
      where: { id },
    }, info);
  },
};


module.exports = Mutations;
