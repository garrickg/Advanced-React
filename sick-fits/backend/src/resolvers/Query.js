const { forwardTo } = require('prisma-binding');

const { hasPermission } = require('../utils');

const Query = {
  items: forwardTo('db'),
  item: forwardTo('db'),
  itemsConnection: forwardTo('db'),
  me(parent, args, context, info) {
    // check if there is a current userId
    if (!context.request.userId) return null;
    return context.db.query.user(
      {
        where: { id: context.request.userId },
      },
      info,
    );
  },
  async users(parent, args, context, info) {
    const { user, userId } = context.request;
    // check if they are logged in
    if (!userId) {
      throw new Error('Please login!');
    }
    // check if user has permission to query users
    hasPermission(user, ['ADMIN', 'PERMISSIONUPDATE']);
    // query all users
    return context.db.query.users({}, info);
  },
  async order(parent, { id }, context, info) {
    const { userId, user: { permissions } } = context.request;
    // check if they are logged in
    if (!userId) {
      throw new Error('Please login!');
    }
    // query order
    const order = await context.db.query.order({
      where: { id },
    }, info);
    // check if they have permissions to view order
    const ownsOrder = order.user.id === userId;
    const hasPermissionToViewOrder = permissions.includes('ADMIN');
    if (!ownsOrder || !hasPermissionToViewOrder) {
      throw new Error('Access Denied!');
    }
    // return the order
    return order;
  },
  async orders(parent, args, context, info) {
    const { userId } = context.request;
    if (!userId) {
      throw new Error('Please log in!');
    }
    return context.db.query.orders({
      where: {
        user: { id: userId },
      },
    }, info);
  },
};

module.exports = Query;
