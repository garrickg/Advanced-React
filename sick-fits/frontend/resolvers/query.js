import gql from 'graphql-tag';

import { perPage } from '../config';

const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`;

const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
    items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const ALL_USERS_QUERY = gql`
  query {
    users {
      id
      name
      email
      permissions
    }
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      largeImage
      price
    }
  }
`;

const CURRENT_USER_QUERY = gql`
  query {
    me {
      id
      email
      name
      permissions
    }
  }
`;

export {
  LOCAL_STATE_QUERY,
  ALL_ITEMS_QUERY,
  PAGINATION_QUERY,
  ALL_USERS_QUERY,
  SINGLE_ITEM_QUERY,
  CURRENT_USER_QUERY,
};
