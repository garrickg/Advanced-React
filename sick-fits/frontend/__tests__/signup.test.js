import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { ApolloConsumer } from 'react-apollo';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';

import Signup from '../components/signup';
import { fakeUser } from '../lib/testUtils';
import { SIGNUP_MUTATION } from '../resolvers/mutation';
import { CURRENT_USER_QUERY } from '../resolvers/query';

function type(wrapper, name, value) {
  wrapper.find(`input[name="${name}"]`).simulate('change', {
    target: { name, value },
  });
}

const me = fakeUser();

const mocks = [
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        email: me.email,
        name: me.name,
        password: 'password',
      },
    },
    result: {
      data: {
        signup: {
          __typename: me.__typename,
          id: me.id,
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: { me },
    },
  },
];

describe('<Signup/>', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <Signup />
      </MockedProvider>
    );
    expect(toJSON(wrapper.find('form[data-test="form"]'))).toMatchSnapshot();
  });
  it('calls the mutation properly', async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {(client) => {
            apolloClient = client;
            return <Signup />;
          }}
        </ApolloConsumer>
      </MockedProvider>
    );
    await wait();
    type(wrapper, 'name', me.name);
    type(wrapper, 'email', me.email);
    type(wrapper, 'password', 'password');
    wrapper.update();
    wrapper.find('form[data-test="form"]').simulate('submit');
    await wait();
    // query the user out of the apollo client
    const user = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(user.data.me).toMatchObject(me);
  });
});
