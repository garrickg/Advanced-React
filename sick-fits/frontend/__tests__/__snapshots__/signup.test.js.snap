// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="form"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign up for an account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        required={true}
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="name"
    >
      Name
      <input
        name="name"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up!
    </button>
  </fieldset>
</form>
`;
